torch/csrc/deploy/builtin_registry.cpp

```diff
@@ -0,0 +1,85 @@
+#include <Python.h>
+#include <torch/csrc/deploy/builtin_registry.h>
+
+namespace torch {
+namespace deploy {
+
+builtin_registry_item::builtin_registry_item(
+    const char* _name,
+    const struct _frozen* _frozen_modules)
+    : name(_name), frozen_modules(_frozen_modules) {
+  num_modules = 0;
+  if (frozen_modules) {
+    while (frozen_modules[num_modules].name != nullptr) {
+      ++num_modules;
+    }
+  }
+
+  fprintf(
+      stderr,
+      "torch::deploy builtin %s contains %d modules\n",
+      name,
+      num_modules);
+}
+
+builtin_registry* builtin_registry::get() {
+  static builtin_registry _registry;
+  return &_registry;
+}
+
+void builtin_registry::register_builtin(
```

Comment:
yes, it's safer to add mutex to guard the accesses. But giving it another think, I feel maybe we don't need that?

Typical use of the BuilitinRegistry is we call macro REGISTER_TORCH_DEPLOY_BUILTIN in the global scope of a source file. The macro just create an anonymous registerer object. The constructor of those registerer object will call builtin_registry::register_builtin. Those calls to the registerer's constructor will be collected to .init section of the embedded interpreter and being executed when we call dlopen.

So I think for this typical scenario, things are done in a single thread. Adding mutex for all accesses will make the code a bit complex.

Let me know if miss something

Reply:
this is true if we assume that the macro will only ever be called during static initialization, but it's not necessarily the case. Our experience with custom operators (which are registered in a similar way) suggests that sometimes people call them during runtime as well. Since we don't control it, better to be safe I think?

Reply:
> or hmmm, I guess since they have to be part of the embedded interpreter build it doesn't make sense for them to be called outside of static initialization? Is that what you're saying?

yep. Also there is very little chance that people can register a torch::deploy builtin at runtime. Since if they really want to do that, they need to do that before we iterate thru the list of registered builtins during initialization in interpreter_impl.cpp.